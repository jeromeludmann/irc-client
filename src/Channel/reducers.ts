import { combineReducers } from "redux";
import {
  CommandInputState,
  commandInputReducer,
} from "@app/CommandInput/reducers";
import { ChannelScope } from "@app/types";

// Name

type NameState = string;
const nameInitialState: NameState = "";
const nameReducer = (state = nameInitialState): NameState => state;

// Messages

type MessagesState = string[];

type MessagesPayload = ChannelScope & { value: string };

interface MessagesAction {
  type: "COMMAND_SENT";
  payload: MessagesPayload;
}

const messagesInitialState: MessagesState = [];

const messagesReducer = (
  state = messagesInitialState,
  { type, payload }: MessagesAction,
): MessagesState => {
  switch (type) {
    case "COMMAND_SENT":
      return [...state, payload.value];
    default:
      return state;
  }
};

export interface ChannelState {
  name: NameState;
  messages: MessagesState;
  commandInput: CommandInputState;
}

export const channelReducer = combineReducers<ChannelState>({
  name: nameReducer,
  messages: messagesReducer,
  commandInput: commandInputReducer,
});
