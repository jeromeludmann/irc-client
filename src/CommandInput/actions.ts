import { ChannelScope } from "@app/types";

export const updateCommand = (
  command: string,
  { server, channel }: ChannelScope,
) => ({
  type: "COMMAND_CHANGED",
  payload: { server, channel, value: command },
});

export const sendCommand = (
  command: string,
  { server, channel }: ChannelScope,
) => ({
  type: "COMMAND_SENT",
  payload: { server, channel, value: command },
});
