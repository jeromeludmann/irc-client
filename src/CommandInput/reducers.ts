import { combineReducers } from "redux";

// Value

type ValueState = string;

type ValueAction =
  | {
      type: "COMMAND_CHANGED";
      payload: { value: string };
    }
  | {
      type: "COMMAND_SENT";
      payload: {};
    };

const valueInitialState: ValueState = "";

const value = (state = valueInitialState, action: ValueAction): ValueState => {
  switch (action.type) {
    case "COMMAND_CHANGED":
      return action.payload.value;
    case "COMMAND_SENT":
      return "";
    default:
      return state;
  }
};

// History

type HistoryState = string[];
const historyInitialState: HistoryState = [];
const history = (state = historyInitialState): HistoryState => state;

// CommandInput

export type CommandInputState = {
  value: ValueState;
  history: HistoryState;
};

export const commandInputReducer = combineReducers<CommandInputState>({
  value,
  history,
});
