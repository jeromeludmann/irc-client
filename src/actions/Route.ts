import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export type RoutedAction = Action & {
  route: Route;
};
