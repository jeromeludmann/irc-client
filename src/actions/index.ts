export const SET_ACTIVE_WINDOW = "SET_ACTIVE_WINDOW";

export interface SetActiveWindow {
  type: typeof SET_ACTIVE_WINDOW;
  payload: { server: string; channel: string };
}

export function setActiveWindow(server: string, channel: string) {
  return {
    type: SET_ACTIVE_WINDOW,
    payload: { server, channel },
  };
}
