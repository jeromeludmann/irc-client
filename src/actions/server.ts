import { Action } from "redux";

export const CONNECT_SERVER = "SERVER/CONNECT_SERVER";
export const DISCONNECT_SERVER = "SERVER/DISCONNECT_SERVER";

export type ConnectServerAction = Action<typeof CONNECT_SERVER> & {
  payload: {
    host: string;
    port: number;
  };
};

export type DisconnectServerAction = Action<typeof DISCONNECT_SERVER>;

export const connectServer = (
  host: string,
  port: number = 6667,
): ConnectServerAction => ({
  type: CONNECT_SERVER,
  payload: { host, port },
});

export const disconnectServer = (): DisconnectServerAction => ({
  type: DISCONNECT_SERVER,
});
