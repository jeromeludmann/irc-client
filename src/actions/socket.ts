import { Action } from "redux";

export const LOOKUP_SUCCESS = "SOCKET/LOOKUP_SUCCESS";
export const LOOKUP_FAILED = "SOCKET/LOOKUP_FAILED";
export const MESSAGES_RECEIVED = "SOCKET/RECEIVE_MESSAGES";
export const CONNECTION_ESTABLISHED = "SOCKET/CONNECTION_ESTABLISHED";
export const CONNECTION_CLOSED = "SOCKET/CONNECTION_CLOSED";
export const CONNECTION_FAILED = "SOCKET/CONNECTION_FAILED";

export type LookupSuccessAction = Action<typeof LOOKUP_SUCCESS> & {
  address: string;
  family: string | number | null;
  host: string;
};

export type LookupFailedAction = Action<typeof LOOKUP_FAILED> & {
  error: Error;
};

export type MessagesReceivedAction = Action<typeof MESSAGES_RECEIVED> & {
  payload: {
    serverId: string;
    messages: string[];
  };
};

export type ConnectionEstablishedAction = Action<typeof CONNECTION_ESTABLISHED>;

export type ConnectionClosedAction = Action<typeof CONNECTION_CLOSED> & {
  payload: {
    hadError: boolean;
  };
};

export type ConnectionFailedAction = Action<typeof CONNECTION_FAILED> & {
  payload: {
    name: string;
    message: string;
  };
};

export const lookup = (
  error: Error | null,
  address: string,
  family: string | number | null,
  host: string,
): LookupSuccessAction | LookupFailedAction => {
  return error
    ? { type: LOOKUP_FAILED, error }
    : { type: LOOKUP_SUCCESS, address, family, host };
};

export const setMessagesReceived = (
  serverId: string,
  messages: string[],
): MessagesReceivedAction => ({
  type: MESSAGES_RECEIVED,
  payload: { serverId, messages },
});

export const setConnectionEstablished = (): ConnectionEstablishedAction => ({
  type: CONNECTION_ESTABLISHED,
});

export const setConnectionClosed = (
  hadError: boolean,
): ConnectionClosedAction => ({
  type: CONNECTION_CLOSED,
  payload: { hadError },
});

export const setConnectionFailed = (
  name: string,
  message: string,
): ConnectionFailedAction => ({
  type: CONNECTION_FAILED,
  payload: { name, message },
});
