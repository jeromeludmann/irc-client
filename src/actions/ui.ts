import { Action } from "redux";
import { Route } from "@app/Route";

// Update input value

export const UPDATE_INPUT_VALUE = "UI/UPDATE_INPUT_VALUE";

export interface UpdateInputValueAction
  extends Action<typeof UPDATE_INPUT_VALUE> {
  payload: { value: string };
}

export function updateInputValue(value: string): UpdateInputValueAction {
  return { type: UPDATE_INPUT_VALUE, payload: { value } };
}

// Enter input value

export const ENTER_INPUT_VALUE = "UI/ENTER_INPUT_VALUE";

export interface EnterInputValueAction
  extends Action<typeof ENTER_INPUT_VALUE> {
  payload: { value: string };
}

export function enterInputValue(value: string): EnterInputValueAction {
  return { type: ENTER_INPUT_VALUE, payload: { value } };
}

// Go back input history

export const GO_BACK_INPUT_HISTORY = "UI/GO_BACK_INPUT_HISTORY";

export interface GoBackInputHistoryAction
  extends Action<typeof GO_BACK_INPUT_HISTORY> {}

export function goBackInputHistory(): GoBackInputHistoryAction {
  return { type: GO_BACK_INPUT_HISTORY };
}

// Go forward input history

export const GO_FORWARD_INPUT_HISTORY = "UI/GO_FORWARD_INPUT_HISTORY";

export interface GoForwardInputHistoryAction
  extends Action<typeof GO_FORWARD_INPUT_HISTORY> {}

export function goForwardInputHistory(): GoForwardInputHistoryAction {
  return { type: GO_FORWARD_INPUT_HISTORY };
}

// Switch route

export const SWITCH_WINDOW = "UI/SWITCH_WINDOW";

export interface SwitchWindowAction extends Action<typeof SWITCH_WINDOW> {
  route: Route;
}

export function switchWindow(route: Route): SwitchWindowAction {
  return {
    type: SWITCH_WINDOW,
    route,
  };
}

// Close window

export const CLOSE_WINDOW = "UI/CLOSE_WINDOW";

export interface CloseWindowAction extends Action<typeof CLOSE_WINDOW> {
  route: Route;
}

export function closeWindow(route: Route): CloseWindowAction {
  return {
    type: CLOSE_WINDOW,
    route,
  };
}
