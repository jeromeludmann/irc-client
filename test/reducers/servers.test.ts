import { reduceServers, serversInitialState } from "@app/reducers/servers";
import { CLOSE_WINDOW } from "@app/actions/ui";
import { STATUS } from "@app/Route";
import { serverInitialState } from "@app/reducers/server";
import { LOOKUP_SUCCESS } from "@app/actions/socket";

describe("servers reducer", () => {
  const route = { serverKey: "", channelKey: "" };

  test("initializing without state", () => {
    expect(
      reduceServers(undefined, { type: "", route }, { route }),
    ).toMatchSnapshot();
  });

  test("initializing with state", () => {
    expect(
      reduceServers(serversInitialState, { type: "", route }, { route }),
    ).toMatchSnapshot();
  });

  describe("close window", () => {
    test("without server", () => {
      expect(
        reduceServers(
          serversInitialState,
          { type: CLOSE_WINDOW, route },
          { route },
        ),
      );
    });

    test("close non status window, with two servers", () => {
      expect(
        reduceServers(
          serversInitialState,
          { type: CLOSE_WINDOW, route },
          { route },
        ),
      );
    });

    test("close status window, with two servers", () => {
      expect(
        reduceServers(
          serversInitialState,
          { type: CLOSE_WINDOW, route: { serverKey: "", channelKey: STATUS } },
          { route },
        ),
      );

      expect(
        reduceServers(
          {
            server1: serverInitialState,
            server2: serverInitialState,
          },
          {
            type: CLOSE_WINDOW,
            route: { serverKey: "server1", channelKey: STATUS },
          },
          { route },
        ),
      );
    });

    expect(
      reduceServers(
        { server1: serverInitialState },
        {
          type: LOOKUP_SUCCESS,
          route: { serverKey: "server1", channelKey: "" },
        },
        { route },
      ),
    );
  });
});
