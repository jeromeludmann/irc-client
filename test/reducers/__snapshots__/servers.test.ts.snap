// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`servers reducer initializing with state 1`] = `Object {}`;

exports[`servers reducer initializing without state 1`] = `Object {}`;
